### PR TITLE
Add support for getaddrinfo

### PR DESCRIPTION
**Pull request by bikallem**
Fixes part of https://github.com/ocaml-multicore/eio/issues/258

### CODE REVIEW

File: lib_eio/net.mli [hunk 1]
```diff
@@ -171,6 +172,12 @@ val recv : #datagram_socket -> Cstruct.t -> Sockaddr.datagram * int
     returned along with the sender address and port. If the [buf] is too small then excess bytes may be discarded
     depending on the type of the socket the message is received from. *)
 
+(** {2 DNS queries} *)
+
+val addrinfo: #t -> string -> ([`Tcp | `Udp] * Ipaddr.v4v6) list
```
Comment (talex5) [inline lib_eio/net.mli:177]:
```suggestion
val addrinfo: #t -> string -> Sockaddr.t list
```
(or at least a compatible subset, if you want to exclude Unix-domain sockets)

**Comment by bikallem** [comment]
@talex5 do you know if the `ocaml-ci` has ipv6 enabled? or how to enable it? The test for this PR seems to be failing as I expect Ipv6 address but the machine is not returning any.

File: lib_eio_luv/eio_luv.ml [hunk 2]
```diff
@@ -407,6 +407,24 @@ module Low_level = struct
...
+              let st = match socktype' with `STREAM -> `Tcp | `DGRAM -> `Udp in
+              let ipaddr = Eio_unix.Ipaddr.of_unix (Unix.inet_addr_of_string host) in
+              (st, ipaddr)
+            )
+      | _ -> None
+    in
+    let request = Luv.DNS.Addr_info.Request.make () in
+    await_with_cancel ~request (fun loop -> Luv.DNS.getaddrinfo ~loop ~request ~node:hostname ())
+    |> Result.map (List.filter_map convert_addrinfo)
+    |> or_raise
```
Comment (talex5) [inline lib_eio_luv/eio_luv.ml:426]:
```suggestion
    |> or_raise
    |> List.filter_map convert_addrinfo
```
(just a bit simpler)

File: tests/network.md [hunk 3]
```diff
@@ -485,3 +485,24 @@ EPIPE:
     (fun () -> Eio.Flow.shutdown a `Receive);;
 - : unit = ()
 ```
+
+## Addrinfo (Linx and Libuv)
```
Comment (talex5) [inline tests/network.md:489]:
```suggestion
## Addrinfo (Linux and Libuv)
```

File: tests/network.md [hunk 4]
```diff
@@ -485,3 +485,24 @@ EPIPE:
...
+```ocaml
+# Eio_main.run @@ fun env ->
+  Eio.Net.addrinfo env#net "localhost";;
+- : ([ `Tcp | `Udp ] * Eio.Net.Ipaddr.v4v6) list =
+[(`Tcp, "\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\001");
+ (`Udp, "\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\001");
+ (`Tcp, "\127\000\000\001"); (`Udp, "\127\000\000\001")]
+```
+
+<!-- $MDX set-EIO_BACKEND=luv -->
```
Comment (talex5) [inline tests/network.md:500]: We need a proper fix for this (at the moment, I run `make test_luv` to run all the tests this way). Probably we should have a `tests/luv` directory with some dune magic to duplicate all the tests.

**Review by talex5 — COMMENTED** [review]
Thanks! This looks reasonable to me.

I think we should call this `getaddrinfo` rather than `addrinfo` though, as that's the name everyone uses for it (including OCaml's Unix module). Also, it may access the network so the getting part is quite important.

> do you know if the ocaml-ci has ipv6 enabled? or how to enable it? The test for this PR seems to be failing as I expect Ipv6 address but the machine is not returning any.

What gets returned for `localhost` will depend on the system configuration, so it's probably best to test it using numeric addresses.

File: lib_eio/net.mli (hunk 1 above)
Comment (bikallem) [reply lib_eio/net.mli]: Ended up using `Sockaddr.t` as suggested.

File: lib_eio_luv/eio_luv.ml (hunk 2 above)
Comment (bikallem) [reply lib_eio_luv/eio_luv.ml]: Done.

File: tests/network.md (hunk 3 above)
Comment (bikallem) [reply tests/network.md]: Removed the redundant line and luv test.

File: tests/network.md (hunk 4 above)
Comment (bikallem) [reply tests/network.md]: Removed luv test and using `make test_luv` instead.

**Comment by bikallem** [comment]
@talex5 I have renamed `addrinfo` to `getaddrinfo` as suggested and removed the luv specific test from `network.md`.

**Comment by SquidDev** [comment]
One thing possibly worth mentioning is that users may wish to override this method to point to a non-libc implementation (especially on Alpine, where musl's `getaddrinfo` implementation is intentionally limited).

Supporting this use case isn't a problem (capabilities make this very easy to do), but I do feel a bit weird about calling this function `getaddrinfo` when it no longer uses the POSIX function - I'd personally go for something like `resolve_addr`.

**Comment by bikallem** [comment]
@SquidDev Thanks for the feedback. Don't really have any personal preference but I agree with @talex5 - `getaddrinfo` is the more common term for what we are doing here since both sets of backend (luv and unix) both name it as such.

**Review by talex5 — COMMENTED** [review]
Either `getaddrinfo` or `resolve_addr` is fine with me. I don't think there's any problem with `getaddrinfo` calling an alternative implementation as long as it does roughly the same thing. Otherwise we'd have to rename everything else too (e.g. `unlink` and `mkdir` might not be calling the glibc function either, if it's a remote directory, or a directory inside a zip archive).

File: lib_eio/net.mli [hunk 5]
```diff
@@ -171,6 +172,12 @@ val recv : #datagram_socket -> Cstruct.t -> Sockaddr.datagram * int
     returned along with the sender address and port. If the [buf] is too small then excess bytes may be discarded
     depending on the type of the socket the message is received from. *)
 
+(** {2 DNS queries} *)
+
+val getaddrinfo: #t -> string -> Sockaddr.t list
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply lib_eio/net.mli:177]: I'd like that !

File: lib_eio/mock/net.ml [hunk 6]
```diff
@@ -5,18 +5,21 @@ type t = <
...
   on_connect : <Eio.Net.stream_socket; Eio.Flow.close> Handler.t;
   on_datagram_socket : <Eio.Net.datagram_socket; Eio.Flow.close> Handler.t;
+  on_getaddrinfo : Eio.Net.Sockaddr.t list Handler.t;
 >
 
 let make label =
   let on_listen = Handler.make (`Raise (Failure "Mock listen handler not configured")) in
   let on_connect = Handler.make (`Raise (Failure "Mock connect handler not configured")) in
   let on_datagram_socket = Handler.make (`Raise (Failure "Mock datagram_socket handler not configured")) in
+  let on_getaddrinfo = Handler.make (`Raise (Failure "Mock addrinfo handler not configured")) in
```
[the remark this answers is not in the document]
Comment (bikallem) [reply lib_eio/mock/net.ml:15]: Ack.

File: lib_eio/mock/net.ml [hunk 7]
```diff
@@ -49,6 +56,9 @@ let on_datagram_socket (t:t) actions =
   let as_socket x = (x :> <Eio.Net.datagram_socket; Eio.Flow.close>) in
   Handler.seq t#on_datagram_socket (List.map (Action.map as_socket) actions)
 
+let on_getaddrinfo (t:t) actions =
+  Handler.seq t#on_getaddrinfo (List.map (Action.map Fun.id) actions)
```
[the remark this answers is not in the document]
Comment (bikallem) [reply lib_eio/mock/net.ml:60]: Ack.

File: lib_eio/net.mli (hunk 5 above)
[the remark this answers is not in the document]
Comment (bikallem) [reply lib_eio/net.mli]: Done. I have added the `service` parameter now.

**Review by talex5 — APPROVED** [review]
Looks good to me! I added one test for port numbers and squash-rebased it on `main`.

**Pull request merged**